lib/src/js/compiler.dart

```diff
@@ -0,0 +1,90 @@
+import 'dart:js_util';
+
+import 'package:node_interop/js.dart';
+
+import 'compile.dart';
+import 'compile_options.dart';
+import 'reflection.dart';
+import 'utils.dart';
+
+class Compiler {
+  bool _disposed = false;
+
+  void throwIfDisposed() {
+    if (_disposed) {
+      jsThrow(JsError('Compiler has already been disposed.'));
+    }
+  }
+}
+
+class AsyncCompiler extends Compiler {
+  final Set<Promise> _compilations = {};
+
+  /// Adds a compilation to the pending set and removes it when it's done.
+  void _addCompilation(Promise compilation) {
+    _compilations.add(compilation);
+    compilation.then((value) {
+      _compilations.remove(compilation);
+    }, (error) {
+      _compilations.remove(compilation);
+    });
+  }
+}
+
+/// The JavaScript `Compiler` class.
+final JSClass compilerClass = () {
+  var jsClass = createJSClass('sass.Compiler', () => Compiler());
```

Comment:
I think we should throw an error if the user attempts to manually construct a compiler instance without going through the `init` functions. Can you add that to the spec/tests?

Reply:
@nex3 Looking at the proposal, we should only be exposing the Compiler type, but not the class itself. 

That means the change we need to make is updating the embedded host and dart-sass here to not export the class, and then there wouldn't be a risk of a user calling `new Compiler()`- correct?

Reply:
I see James has already implemented a constructor that throws in https://github.com/sass/dart-sass/pull/2134/commits/53514e390f50cd5bcbf82136222cce8946a76253. The Dart codebase uses that trick in several places (and I used it too to make SassCalculations throw), but I wonder how we should implement this in the embedded host.

We can update the spec to mark the constructor as private so tsc yells at authors for constructing Compiler instances, but this won't stop anyone from actually doing it. In fact, this is the case for [`SassCalculation`](https://github.com/sass/embedded-host-node/blob/06974e1f0fdded65af93135fb1ea39fcbec2c2eb/lib/src/value/calculations.ts#L34). I just checked and I can instantiate it in plain JS because there's no runtime check, which is an inconsistency with the dart-sass implementation.

This raises the bigger picture question of why do we want to use the `init*` functions and disallow direct instantiation? I think enforcing this will require some closure tricks (which is what Dart does on the code it generates), unless I'm missing a more straight-forward approach.

Reply:
Clarification: I had forgotten Natalie mentioned that the types used by the embedded host are not consumed by end users, so the fact that the constructor is private there is not relevant unless it's also private in the spec.

Reply:
> This raises the bigger picture question of why do we want to use the `init*` functions and disallow direct instantiation?

I went through the same thought process- the crucial point is that `initAsyncCompiler` returns a Promise, and constructors can't be async.

Because the types are defined in the sass pkg, in node-embedded-host I was able to add a private flag to the constructor that can only be passed via the init function - https://github.com/sass/embedded-host-node/pull/261/commits/7b7fbf7b01d2850a2a9353630ffcc21768719fab, allowing us to avoid the IIFE alternatives.